### PR TITLE
fix: update typescript and typescript-estree to latest

### PR DESCRIPTION
**Pull request by JamesHenry**
This updates `typescript` and `typescript-estree` to latest.

I have made some progress on the breaking changes, but I was not able to finish before breaking off, so I have committed the outstanding regressions in a second commit so that they are easier for someone else to diagnose and/or fix if they have the bandwidth. I will likely not be able to pick this up again until next weekend, personally.

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

Fixes https://github.com/prettier/prettier/issues/2347

### CODE REVIEW

**Comment by JamesHenry** [comment]
Found a little time so fixed up some more issues and redid the regressions commit

**Comment by JamesHenry** [comment]
Fixed up everything except for this: https://github.com/JamesHenry/typescript-estree/issues/109

**Comment by JamesHenry** [comment]
Unit tests now passing, but apparently not with AST_COMPARE enabled... :/

**Comment by JamesHenry** [comment]
Ah... The issue is that the TypeScript AST contains `TSParenthesizedType` when the type is wrapped in parens - i.e. the AST will be fundamentally different if prettier adds or removes parens from a type.

To be honest, I'm not sure why this wasn't an issue before. We weren't explicitly converting `TSParenthesizedType` before, but I would have thought it would have been picked up by our auto-conversion fallback logic...

Any thoughts/ideas @j-f1 @ikatyang @vjeux?

**Comment by JamesHenry** [comment]
Ah sorry I think I've found it, clean.js needs to be updated - testing now...

**Comment by JamesHenry** [comment]
Woah - the build is failing because Prettier is being tested in a Node 4 environment!

Node v4 hit end of life on `2018-04-30` - can we remove that environment now?

https://github.com/nodejs/Release#end-of-life-releases

**Comment by ikatyang** [comment]
> Node v4 hit end of life on 2018-04-30 - can we remove that environment now?

That'd be a breaking change so we need to wait until v2.

**Comment by JamesHenry** [comment]
Ok thanks for the inventive monkey patching @ikatyang - it's nice that I don't have to _introduce_ node 4 support to typescript-estree to get around this :)

I am just addressing the coverage stat and updating to the very latest typescript-estree and then this should be good to go!

**Comment by JamesHenry** [comment]
Ready for review!

**Comment by JamesHenry** [comment]
Thanks @duailibe - I'll make sure @ikatyang also gets chance to give a final review before merging

**Comment by ikatyang** [comment]
Can you add a test case for #2347?

**Comment by JamesHenry** [comment]
Good point - I'll have to add it later tonight or tomorrow, when I am back on my personal machine

File: src/language-js/printer-estree.js [hunk 1]
```diff
@@ -772,13 +772,7 @@ function printPathNoParens(path, options, print, args) {
       ]);
     case "FunctionDeclaration":
     case "FunctionExpression":
-      if (isNodeStartingWithDeclare(n, options)) {
```
[the remark this answers is not in the document]
Comment (armano2) [reply src/language-js/printer-estree.js]: `FunctionDeclaration` without body is now `TSDeclareFunction` and 
`declare function foo();`
is now `TSDeclareFunction[declare=true]`

[the remark this answers is not in the document]
Comment (JamesHenry) [reply src/language-js/printer-estree.js]: Yes exactly, I was just pointing out to the reviewers why this was safe to remove 👍

File: src/language-js/needs-parens.js [hunk 2]
```diff
@@ -334,12 +332,12 @@ function needsParens(path, options) {
       if (
         (parent.type === "TSTypeParameter" ||
           parent.type === "TypeParameter" ||
-          parent.type === "VariableDeclarator" ||
+          parent.type === "TSTypeAliasDeclaration" ||
           parent.type === "TSTypeAnnotation" ||
+          parent.type === "TSParenthesizedType" ||
           parent.type === "GenericTypeAnnotation" ||
```
Comment (armano2) [inline src/language-js/needs-parens.js]: `GenericTypeAnnotation` can't be produced by estree anymore it's now always `TSTypeAnnotation`

File: src/language-js/printer-estree.js [hunk 3]
```diff
@@ -1706,6 +1690,29 @@ function printPathNoParens(path, options, print, args) {
...
+        " =",
+        printed,
+        semi
+      );
+
+      return group(concat(parts));
+    }
     case "VariableDeclarator":
       return printAssignment(
         n.id,
```
Comment (armano2) [inline src/language-js/printer-estree.js]: VariableDeclarator will no longer have `typeParameters`

File: src/language-js/printer-estree.js [hunk 4]
```diff
@@ -772,13 +772,7 @@ function printPathNoParens(path, options, print, args) {
       ]);
     case "FunctionDeclaration":
     case "FunctionExpression":
-      if (isNodeStartingWithDeclare(n, options)) {
-        parts.push("declare ");
-      }
       parts.push(printFunctionDeclaration(path, print, options));
-      if (!n.body) {
```
[the remark this answers is not in the document]
Comment (armano2) [reply src/language-js/printer-estree.js]: that's not true, it's still possible for `FunctionExpression`

```
abstract class AbstractSocket {
    createSocket?(): Promise<string>;
}
```

https://github.com/armano2/typescript-estree-test/blob/master/typescript-estree.types.d.ts#L326

**Comment by JamesHenry** [comment]
@ikatyang should be good to go now! Feel free to merge if you are happy

**Comment by armano2** [comment]
i'm unsure if this should be solved here, or as separate PR, 

```
type Unpacked<T> =
  T extends (infer U)[] ? U :
    T extends infer U ? U :
      T extends Promise<infer U> ? U :
        T;
```
will be formatted to unreadable

```
type Unpacked<T> = T extends (infer U)[]
  ? U
  : T extends infer U
  ? U
  : T extends Promise<infer U>
  ? U
  : T;
```

**Pull request merged**